src/hammer-vlsi/synthesis/vivado/core.py

```diff
@@ -0,0 +1,80 @@
+from typing import Dict, List
+
+import os
+import errno
+import pkg_resources
+from itertools import chain
+
+from jinja2 import Template
+
+from hammer_vlsi import deepdict
+
+
+class VivadoCommon(object):
+    @property
+    def env_vars(self) -> Dict[str, str]:
+        new_dict = deepdict(super().env_vars)
+        return new_dict
+
+    def sym_link_force(self, source: str, dest: str) -> None:
+        try:
+            os.symlink(source, dest)
+        except OSError as e:
+            if e.errno == errno.EEXIST:
+                os.remove(dest)
+                os.symlink(source, dest)
+            else:
+                raise e
+
+    def setup_workspace(self, add_files: List[str], run_script: str,
+                        extra_params: Dict[str, str]) -> None:
+        # make working directory
+        os.makedirs(self.run_dir, exist_ok=True)
+        os.makedirs(os.path.join(self.run_dir, 'obj'), exist_ok=True)
+        # create symlink to constraint file
+        cons_fname = os.path.abspath(
+            self.get_setting('synthesis.vivado.constraints_file'))
+        cons_dir = os.path.join(self.run_dir, 'constrs')
+        os.makedirs(cons_dir, exist_ok=True)
+        cons_targ = os.path.join(cons_dir, os.path.basename(cons_fname))
+        self.sym_link_force(cons_fname, cons_targ)
```

Comment:
I think it the constraints file isn't going to be that big, we may as well copy it to be deterministic.